### PR TITLE
Add tools to Spotify Toolkit

### PR DESCRIPTION
**Pull request by EricGustin**
# PR Description
1. `adjust_playback_position` - Adjust the playback position within the currently playing track
2. `skip_to_previous_track` - Skip to the previous track in the user's queue, if any
3. `skip_to_next_track` - Skip to the next track in the user's queue, if any
4. `pause_playback` - Pause the currently playing track, if any
5. `resume_playback` - Resume the currently playing track, if any
6. `start_tracks_playback_by_id` - Start playback of a list of tracks (songs)
7. `get_playback_state` - Get information about the user's current playback state, including track or episode, and active device
8. `get_currently_playing` - Get information about the user's currently playing track
9. `get_track_from_id` - Get information about a track
10. `get_recommendations` - Get track (song) recommendations based on seed artists, genres, and tracks, and multiple target audio stats
11. `get_tracks_audio_features` - Get audio features for a list of tracks (songs)
----------------------

My favorite feature of this toolkit is
1. Start playing my favorite song
2. Get the song that I'm currently playing
3. Get audio features of that song
4. Ask for recommended songs that are similar to it
5. Jam out

------------

### CODE REVIEW

File: toolkits/spotify/arcade_spotify/tools/player.py [hunk 1]
```diff
@@ -1,162 +1,221 @@
...
 
-    Raises:
-        ToolExecutionError: If the request fails for any reason.
+    Note:
+        Either absolute_position_ms or relative_position_ms must be provided, but not both.
     """
-    url = f"{SPOTIFY_BASE_URL}{endpoint}"
-    headers = {"Authorization": f"Bearer {context.authorization.token}"}
+    if (absolute_position_ms is None) == (relative_position_ms is None):
+        raise ValueError(
```
Comment (nbarbettini) [inline toolkits/spotify/arcade_spotify/tools/player.py:39]: Consider a RetryableToolError here?

File: toolkits/spotify/arcade_spotify/tools/tracks.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
+from typing import Annotated, Optional
```
Comment (EricGustin) [inline toolkits/spotify/arcade_spotify/tools/tracks.py:1]: Still need to flesh out the tools in `tracks.py`

File: toolkits/spotify/arcade_spotify/tools/utils.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    """
+    headers = {"Authorization": f"Bearer {context.authorization.token}"}
+
+    async with httpx.AsyncClient() as client:
+        response = await client.request(method, url, headers=headers, params=params, json=json_data)
+
+    return response
+
+
+def handle_spotify_response(response: httpx.Response, url: str):
```
Comment (nbarbettini) [inline toolkits/spotify/arcade_spotify/tools/utils.py:40]: Based on the exception handling Sam put in recently, I think this kind of method is no longer needed at all. Each tool could simply call `response.raise_for_status()` -- see https://www.python-httpx.org/quickstart/#response-status-codes

File: toolkits/spotify/arcade_spotify/tools/player.py (hunk 1 above)
Comment (EricGustin) [reply toolkits/spotify/arcade_spotify/tools/player.py]: Agree, added

File: toolkits/spotify/arcade_spotify/tools/utils.py (hunk 3 above)
Comment (EricGustin) [reply toolkits/spotify/arcade_spotify/tools/utils.py]: Yep, updated

**Review by nbarbettini — APPROVED** [review]
This works great. ✅ 

Some prompts I tried that didn't work:
`play some cool electronic music`
`play me some music by John Williams`

It seems like there isn't a way to directly go from genre/artist -> play track. Would it make sense to have a separate tool for that?

**Comment by EricGustin** [comment]
> The tools in this toolkit work great. ✅
> 
> For discussion - some prompts I tried that didn't work: `play some cool electronic music` `play me some music by John Williams`
> 
> It seems like there isn't a way to directly go from genre/artist -> play track. Would it make sense to have a separate tool for that?

I think by introducing more tools in the future that deal with genres and artists could likely indirectly expand the capabilities of the `get_recommendations` tool

**Pull request merged**